### PR TITLE
perf(data): the streamed resample runs on the device, out of core, bounded by the machine

### PR DESCRIPTION
**Pull request by vboussot**
Stacked on #110 (`feat/config-strict-read`), itself on #109 and #105: merge those first, then retarget this PR to `main`. Ported onto the audit stack (it was written against #105 alone): what both had done independently is the stack's version, the perf work is this commit.

## What (one commit)

- **The streamed resample runs on the device, out of core.** The TRANSFORM coordinate walk is vectorised, slabbed under a budget read off the machine, and runs on the rank's device end to end, reductions included. Measured on a three-specimen ExaSPIM template round, one GPU: Fold 127 s + Update 73 s where the CPU baseline did not finish in 1 h 47; 4.5 GiB peak VRAM; bitwise-deterministic; CPU-vs-GPU within 8e-6 um. The float64 walk stays bit-identical to `sitk.Resample` (pinned by `torch.equal` tests); `Resample` may declare `precision: fast` (float32 walk) for intensity chains.
- **On a machine without a GPU, `Resample` goes through ITK's own resampler** (each streamed region's source window as a sitk image, the stages one sitk transform): 27 ns/voxel where the torch walk on the host does 326. Full CPU Fold on the same round: 2005 s -> 892 s.
- **A pyramid appended to a streamed store no longer rewrites level 0**: coarser levels are computed lazily from it (dask coarsen, no wasm sandbox) and stored beside it; each declared factor shrinks the level above it, as the release's contract says (ngff-zarr's level-0-relative argument is derived). Update.yml on the same round: 103 s -> 53 s on GPU; `write_ome_zarr` creates the store empty and lets zarr fill it (2.3 s where the dask rechunk took 14.4 s for 2.1 GB).
- **Vote** counts along the case axis (2 s per 512³ where a `torch.unique` took 31 s); **Median** reads the middle off a sort along the case axis (bit-identical, 1.5-3.5x). A reduction folds on the rank's device, its regions re-fitted to the card's free memory; the folds a statistics pass computes are kept for the write pass when the folded output fits half the budget.
- **The per-rank thread share sizes ITK's pool with torch's** and counts the cores a process may run on (affinity mask and cgroup `cpu.max`, not the host's `os.cpu_count()`): `--cpu 3` on three 512³ cases 15.7 s -> 9.1 s. Publishing an entry retires the staging debris of dead writers beside it.
- **The write side of the streaming contract is proven the way the read side is**: `tests/unit/test_transform_materialize_contract.py` materializes every built-in transform slab by slab into a `Write` and compares the store with the whole-volume path's, on CPU and CUDA, into mha and omezarr, read from mha, nii, nii.gz, h5.

## Port notes
Dropped as already in the stack: the cgroup budget, byte-balanced shards, lazy `dicom`/`ome_zarr` imports, probe cleanup, `--plan` flags, h5 replace-aside, ITK_ key filter, `Padding` streaming (the stack's handles reflect). The working set stays the stack's estimator (fallback factor + each stage's declared working multiple). Two stray log files the draft carried are gone.

## Verification
Full core suite (unit + integration) green on the ported tree except one integration test that hung under 8 parallel workers (`test_konfai_cli_resume_continues_training`, a CLI training rendezvous under load; passes alone, and it exercises the stack's code, not this commit's); ruff, format, mypy, bandit green; konfai-mcp / Studio / apps suites run before push.

## Summary by CodeRabbit

- **New Features**
  - Added GPU selection and quiet-mode options for transform planning.
  - Added configurable resampling precision, channel expansion, 2D NIfTI streaming, and device-aware reductions.
  - Added faster OME-Zarr chunk reads and pyramid generation.

- **Improvements**
  - Improved CPU/GPU memory budgeting, workload distribution, coordinate accuracy, and integer-label handling.
  - Added atomic file replacement and safer checkpoint publishing.
  - Improved metadata, displacement fields, vector images, and streamed materialization consistency.

- **Documentation**
  - Clarified streaming accuracy, device behavior, workload sharding, budgets, validation, and destination restrictions.

### CODE REVIEW

**Comment by vboussot** [comment]
@coderabbitai review

**Comment by vboussot** [comment]
@coderabbitai review

File: docs/source/concepts/streaming.md [hunk 1]
```diff
@@ -235,14 +231,25 @@ it in torch: same values, different summation order, so a voxel may land a few
...
+integer volumes). Nearest-neighbour, which is what a `uint8` label volume gets,
+picks on the exact index; cubic walks its own corners; an axis-aligned change of
+density is read one axis at a time on global coordinates: all three are
+bit-identical everywhere.
+
+The slab height follows the budget, so it can differ between machines; through
+that same non-separable linear resample two runs of one chain under different
+budgets then differ by the same ~1e-5, and the plan says so when the budget
+lowers the height below the default. Everything else is independent of the
+slabbing: the same chain writes the same bytes under any budget.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/concepts/streaming.md]: Fixed in cc6126f: the configuration guide scopes the byte claim (bit for bit for pointwise/halo/orientation/crop and nearest/cubic/axis-aligned resamples; ~1e-5 for a linear resample through a non-factorising map on CUDA, linking the streaming guide) and both pages say the bounds assume `precision: exact`.

File: docs/source/config_guide/transform.md [hunk 2]
```diff
@@ -42,9 +42,17 @@ default chunking.
...
 ```
 
+`--cpu N` (or `--gpu` with several ids) shards the *work items* over the ranks,
+balanced by bytes: an ordinary chain is one item per case, a reduction one item
+for all of its cases (so it never gains from more ranks). Every rank writes its
+own entries into the same output, which a single-file store cannot serve: a
+`Save`/`Write` to `:h5` refuses more than one rank; use `:omezarr`, `:mha` or
+`:nii`. An explicit `memory_budget` is per rank, and the plan header says what
+that makes on the node.
+
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/config_guide/transform.md]: Fixed in cc6126f: the earlier paragraph is gone, the fuller one stays.

File: konfai/data/reduction.py [hunk 3]
```diff
@@ -182,7 +182,16 @@ class Median(Reduction):
...
+        # without its interpolation machinery around it (1.5-2x on CPU, 3.5x on CUDA over three
+        # cases, measured). The even-count middle is the same lerp quantile uses, so the two are
+        # bit-identical, odd and even.
+        ranked = torch.stack(tensors, dim=0).float().sort(dim=0).values
+        cases = ranked.shape[0]
+        if cases % 2:
+            middle = ranked[cases // 2]
+        else:
+            middle = torch.lerp(ranked[cases // 2 - 1], ranked[cases // 2], 0.5)
         return middle.to(_averaged_dtype(tensors[0].dtype))
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/reduction.py]: Fixed in cc6126f: the stack is cast to the averaged dtype before the sort (a float64 cohort stays float64), the comment says what the sort allocates.

File: konfai/data/transform.py [hunk 4]
```diff
@@ -3261,14 +3380,18 @@ def patch_locality(self, cache_attribute: Attribute) -> PatchLocality:
...
-            .squeeze(0)
+        # Scattered straight into the float32 answer: ``F.one_hot`` builds the classes in int64
+        # first, three times the bytes of the result for the time of a cast (80 GB for 50 classes
+        # of a 512^3 map, where the answer is 27 GB and this holds the answer plus the labels).
+        result = torch.zeros(
+            (int(tensor.shape[0]), self.num_classes, *tensor.shape[1:]), dtype=torch.float32, device=tensor.device
         )
-        return result
+        result.scatter_(1, tensor.to(torch.int64).unsqueeze(1), 1.0)
+        return result.squeeze(0)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/transform.py:3413]: Fixed in cc6126f: the labels are range-checked before `scatter_` and an out-of-range one raises a `TransformError` naming the case, the range and `num_classes`.

File: konfai/utils/budget.py [hunk 5]
```diff
@@ -139,6 +139,47 @@ def _cgroup_paths() -> list[tuple[Path, str, tuple[str, ...]]]:
...
+        cores = len(os.sched_getaffinity(0))
+    except (AttributeError, OSError):
+        cores = os.cpu_count() or 1
+    for directory in _cpu_cgroup_paths():
+        try:
+            quota, period = (directory / "cpu.max").read_text().split()
+        except (OSError, ValueError):
+            continue
+        if quota != "max":
+            cores = min(cores, max(1, -(-int(quota) // int(period))))
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/budget.py]: Fixed in cc6126f: `_cpu_quota` reads v2 `cpu.max` or v1 `cpu.cfs_quota_us`/`cpu.cfs_period_us` per directory, over the ancestors; test `test_available_cpus_reads_a_cgroup_v1_quota`.

File: konfai/utils/budget.py [hunk 6]
```diff
@@ -139,6 +139,47 @@ def _cgroup_paths() -> list[tuple[Path, str, tuple[str, ...]]]:
...
+    return max(1, cores)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/budget.py]: Fixed in cc6126f: parsing and the division sit inside the try, a malformed or zero-period entry is skipped (asserted in the v2 test).

File: konfai/utils/ITK.py [hunk 7]
```diff
@@ -258,6 +258,49 @@ def decode_transform_stages(transform: sitk.Transform) -> SpatialStages:
...
+            field = sitk.Compose(images)
+            members.append(sitk.DisplacementFieldTransform(field))
+        else:
+            members.append(sitk.BSplineTransform(images, int(stage.order)))
+    if len(members) == 1:
+        return members[0]
+    composite = sitk.CompositeTransform(members[0].GetDimension())
+    for member in reversed(members):
+        composite.AddTransform(member)
+    return composite
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/ITK.py:306]: Fixed in cc6126f: an empty stage list raises a `TransformError` that names the identity spelling.

File: tests/unit/test_data_stream.py [hunk 8]
```diff
@@ -452,3 +452,95 @@ def crash(self, *args, **kwargs):
...
+    got, header_got = streamed.read_data("G", "c")
+    np.testing.assert_array_equal(got, expected)
+    for key in ("Origin", "Spacing", "Direction"):
+        np.testing.assert_allclose(header_got.get_np_array(key), header_expected.get_np_array(key), atol=1e-6)
+
+
+def test_replacing_an_h5_entry_keeps_the_old_one_until_the_new_is_in_place(tmp_path, monkeypatch):
+    """A rewrite moves the old entry aside, publishes, then drops it: at no instant is the entry
+    absent from the file, which is what a crash between the two steps used to leave."""
+    import h5py
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_data_stream.py]: Fixed in cc6126f: `pytest.importorskip("h5py")`.

File: tests/unit/test_ome_zarr_pyramid_chunks.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    for window in windows:
+        got, _ = read_ome_zarr_data_slice(store, window, level=0)
+        np.testing.assert_array_equal(got, volume[window])
+    # A tiny cache: nothing fits, every read decodes -- still the same bytes.
+    OZ._CHUNK_CACHE = OZ._DecodedChunkCache(1024)
+    for window in windows:
+        got, _ = read_ome_zarr_data_slice(store, window, level=0)
+        np.testing.assert_array_equal(got, volume[window])
+    OZ._CHUNK_CACHE = None
+    OZ.clear_ome_zarr_cache()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_ome_zarr_pyramid_chunks.py]: Fixed in cc6126f: try/finally around the tiny-cache loop.

File: konfai/data/transform.py [hunk 10]
```diff
@@ -1257,6 +1276,52 @@ def describe(self) -> str:
...
+    reference.SetDirection(np.asarray(region.direction_xyz, dtype=np.float64).ravel().tolist())
+    outputs = []
+    for channel in range(int(payload.shape[0])):
+        image = sitk.GetImageFromArray(np.ascontiguousarray(payload[channel].numpy()))
+        image.SetOrigin(np.asarray(window_origin, dtype=np.float64).tolist())
+        image.SetSpacing(np.asarray(source.spacing_xyz, dtype=np.float64).tolist())
+        image.SetDirection(np.asarray(source.direction_xyz, dtype=np.float64).ravel().tolist())
+        out = sitk.Resample(image, reference, transform, interpolator[mode], float(fill), image.GetPixelID())
+        outputs.append(torch.from_numpy(sitk.GetArrayFromImage(out)))
+    return torch.stack(outputs, dim=0)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/transform.py]: Measured, then fixed and pinned, in 7e64e66.

The host path now blends in the dtype the walk accumulates in (`sampling_dtype`) and lets torch make the final cast, so ITK no longer accumulates an integer payload in double and casts inside the filter. Both sides then truncate toward zero on the same value.

What remains is real and is now stated rather than claimed away: through a rotation, on an int16 payload, host and walk differ on **2 voxels in 7560, by one unit** -- the ulp of the blend landing on a truncation boundary. `test_the_host_resampler_and_the_walk_agree_on_an_integer_payload` pins exactly that: `nearest` must be exact (tolerance 0), `linear` at most one unit on at most 1/500 of the voxels, for a Euler and an affine map. A real divergence moves voxels by orders of magnitude more and cannot hide under it. The comment at the call site carries the same numbers.

Dropping the host path for integers was the alternative; it would have cost the CPU fold this PR is for (the ExaSPIM round is uint16: 2005 s -> 892 s).

File: konfai/utils/budget.py [hunk 11]
```diff
@@ -139,6 +140,64 @@ def _cgroup_paths() -> list[tuple[Path, str, tuple[str, ...]]]:
...
+            continue
+        hierarchy, controllers, path = parts
+        if hierarchy == "0" and controllers == "":
+            base, v2 = root, True
+        elif "cpu" in controllers.split(","):
+            base, v2 = root / "cpu", False
+        else:
+            continue
+        leaf = base / path.lstrip("/")
+        return [(d, v2) for d in (leaf, *leaf.parents) if d == base or base in d.parents]
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/budget.py:168]: Fixed in 7e64e66: the v1 controller is resolved under the name it is mounted with, `next((root / d for d in (controllers, "cpu") if (root / d).is_dir()), None)`, so `cpu,cpuacct` works whether or not the distribution drops a `cpu` symlink beside it. `test_available_cpus_reads_a_cgroup_v1_quota` is parametrised over both layouts. Reading `/proc/self/mountinfo` would answer the same question with a parser to maintain; the mount name is already in the `/proc/self/cgroup` line.

File: konfai/utils/dataset.py [hunk 12]
```diff
@@ -875,6 +951,8 @@ def _finish(self, success: bool, exc_type, value, traceback) -> None:
         self._finished = True
         try:
             self._close(success)
+            if success and (published := self.published_path) is not None:
+                _retire_dead_debris(published)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/dataset.py]: Fixed in 7e64e66: all three call sites (`DataStream.close`, `SitkFile.data_to_file`, `OmeZarrFile.data_to_file`) run `_retire_dead_debris` under `contextlib.suppress(Exception)` with a comment saying why: it is housekeeping past the publish and cannot fail a write that is complete on disk. `psutil` is a required runtime dependency (`pyproject.toml` line 38, imported at `konfai/__init__.py:24`), not an extra, so no import guard is needed.

File: konfai/utils/dataset.py [hunk 13]
```diff
@@ -1084,7 +1169,13 @@ def __init__(self, path: str, shape: list[int], dtype: np.dtype, attributes: Att
...
             ("BinaryDataByteOrderMSB", "False"),
             ("CompressedData", "False"),
-            ("TransformMatrix", " ".join(str(v) for v in attributes.get_np_array("Direction"))),
+            # MetaIO's TransformMatrix is the TRANSPOSE of ITK's Direction (verified against
+            # sitk.WriteImage): written in Direction order, every non-symmetric orientation reads
+            # back mirrored.
+            (
+                "TransformMatrix",
+                " ".join(str(v) for v in attributes.get_np_array("Direction").reshape(len(spatial), -1).T.ravel()),
+            ),
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/dataset.py:1179]: The regression test already exists and would catch the transpose: `test_stream_matches_whole_volume_write` writes the same volume whole (through `sitk.WriteImage`) and by slabs, then asserts `Direction` equality, and its `_image_attributes()` direction is `[1,0,0, 0,0,-1, 0,1,0]`, which is **not** symmetric (its transpose differs). It runs for `mha` among the four formats.

The compatibility half was missing and is now in 7e64e66: a bug-fix line for the `TransformMatrix` transpose, and a behaviour-change entry saying that an entry streamed to `.mha` by an earlier version with a non-symmetric orientation reads back mirrored, that whole-volume writes and symmetric orientations are unaffected, and how to check and rewrite.

File: konfai/utils/dataset.py [hunk 14]
```diff
@@ -1973,6 +2096,18 @@ def data_to_file(
...
+            try:
+                if final.exists():
+                    final.rename(replaced)
+                staging.rename(final)
+            except BaseException:
+                if replaced.exists() and not final.exists():
+                    replaced.rename(final)
+                raise
+            shutil.rmtree(replaced, ignore_errors=True)
+            _retire_dead_debris(final)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/dataset.py]: Fixed in 7e64e66: the failure branch removes the staging store (`shutil.rmtree(staging, ignore_errors=True)`) before re-raising, matching `_OmeZarrDataStream._close`.
